jhub_apps/service/app_from_git.py

```diff
@@ -0,0 +1,122 @@
+import os
+import pathlib
+import tempfile
+from pathlib import Path
+
+import git
+from fastapi import HTTPException
+from pydantic import ValidationError
+from starlette import status
+
+from jhub_apps.service.models import Repository, AppConfigFromGit
+from jhub_apps.service.utils import logger, encode_file_to_data_url
+
+
+def _clone_repo(repository: Repository, temp_dir):
+    """Clone repository to the given tem_dir"""
+    try:
+        logger.info("Trying to clone repository", repo_url=repository.url)
+        git.Repo.clone_from(repository.url, temp_dir, depth=1, branch=repository.ref)
```

Comment:
> If we are only interested in a single file, should we use git.Repo.archive() with that file name instead?

Not sure, how that would work looking at the docs: https://gitpython.readthedocs.io/en/stable/reference.html?highlight=archive#git.repo.base.Repo.archive

can you elaborate please?

Reply:
See https://graphite.dev/guides/git-clone-single-file#method-2-download-a-single-file-using-git-archive but for gitpython you would use os streams and `tarfile` library. Make sense now or do you want to chat?

Reply:
IIRC, GitHub doesn't supports this (`git archive` API), have you tried this?

Not working(?):
```
$ git archive --format=tar --remote=git@github.com:torvalds/linux.git master README | tar -x

Invalid command: git-upload-archive 'torvalds/linux.git'
  You appear to be using ssh to clone a git:// URL.
  Make sure your core.gitProxy config option and the
  GIT_PROXY_COMMAND environment variable are NOT set.
fatal: the remote end hung up unexpectedly
```

This works on GitLab:

```
git archive --format=tar --remote=git@gitlab.com:to-be-continuous/docker.git master README.md | tar -x
```

Source: https://github.com/isaacs/github/issues/554 (couldn't find official docs on this)

Reply:
Wait, what? GitHub does not support `git` properly, but instead one can use `svn` which is supported by GitHub correctly here? What on earth.

Ref https://stackoverflow.com/questions/9609835/git-export-from-github-remote-repository

Reply:
I think we can move this discussion to a follow-up issue. I think it might be worth implementing if we can easily detect if `git archive` is supported on the remote end.

Reply:
Sounds good, have just created an issue here: https://github.com/nebari-dev/jhub-apps/issues/482